File: BackEnd/Api/Domain/Entities/Dock.cs
namespace Api.Domain.Entities;

using Api.Domain.ValueObjects;
using Api.Application.DataTransfer;
using Api.Infrastructure.Utilities;

public class Dock : IDTOAble<DockDto>
{
    public Guid Id { get; private set; }
    public Code Code { get; private set; }
    public Designation Name { get; private set; }
    public Designation Location { get; private set; }
    public PhysicalCharacteristics PhysicalCharacteristics { get; private set; }
    public virtual ICollection<VesselType> SupportedVesselTypes { get; private set; }


    //EF Core
    protected Dock() { }
    public Dock(Guid id, Code code, Designation name, Designation location, PhysicalCharacteristics physicalCharacteristics, HashSet<VesselType> supportedVesselTypes)
    {
        validatePhysicalCharacteristics(physicalCharacteristics, supportedVesselTypes);

        Id = id;
        Code = code;
        Name = name ?? throw new ArgumentException("Name cannot be null.");
        Location = location ?? throw new ArgumentException("Location cannot be null.");
        PhysicalCharacteristics = physicalCharacteristics;
        SupportedVesselTypes = supportedVesselTypes;
    }

    public void UpdateName(string newName)
    {
        Name = new Designation { Value = newName };
    }
    public void UpdateLocation(string newLocation)
    {
        Location = new Designation { Value = newLocation };
    }
    public void UpdatePhysicalCharacteristics(PhysicalCharacteristics physicalCharacteristics)
    {
        validatePhysicalCharacteristics(physicalCharacteristics, SupportedVesselTypes);
        PhysicalCharacteristics = physicalCharacteristics;
    }

    public void UpdateVesselTypes(ICollection<VesselType> new_vesselTypes)
    {
        validatePhysicalCharacteristics(PhysicalCharacteristics, new_vesselTypes);

        SupportedVesselTypes.Clear();
        SupportedVesselTypes = new_vesselTypes;
    }

    public DockDto ToDTO()
    {
        return new DockDto
        {
            Code = this.Code.Value,
            Name = this.Name.Value,
            Location = this.Location.Value,
            PhysicalCharacteristics = this.PhysicalCharacteristics,
            SupportedVesselTypes = this.SupportedVesselTypes.Select(vt => vt.ToDTO()).ToList()
        };
    }

    private void validatePhysicalCharacteristics(PhysicalCharacteristics physicalCharacteristics, ICollection<VesselType> supportedVesselTypes)
    {
        if (physicalCharacteristics == null)
            throw new ArgumentException("Physical characteristics cannot be null.");
        if (supportedVesselTypes == null || supportedVesselTypes.Count == 0)
            throw new ArgumentException("Invalid vessel types", nameof(supportedVesselTypes));

        foreach (var vt in supportedVesselTypes)
        {
            if (physicalCharacteristics.Length < vt.PhysicalCharacteristics.Length)
                throw new ArgumentException($"The dock's length is insufficient for the vessel type {vt.Name}.");
            if (physicalCharacteristics.Depth < vt.PhysicalCharacteristics.Depth)
                throw new ArgumentException($"The dock's depth is insufficient for the vessel type {vt.Name}.");
            if (physicalCharacteristics.Draft < vt.PhysicalCharacteristics.Draft)
                throw new ArgumentException($"The dock's max draft is insufficient for the vessel type {vt.Name}.");
        }
    }
}

File: FrontEnd/src/model/Dock.ts
import type { DockDto } from "./dto/DockDto";
import type { PhysicalCharacteristics } from "./values/PhysicalCharacteristics";
import type { VesselType } from "./VesselType";

export class Dock {
    private _code: string;
    private _name: string;
    private _location: string;
    private _physicalCharacteristics: PhysicalCharacteristics;
    private _supportedVesselTypes: VesselType[];

    constructor(params: {
        code: string;
        name: string;
        location: string;
        physicalCharacteristics: PhysicalCharacteristics;
        supportedVesselTypes: VesselType[];
    }) {
        this.validatePhysicalCharacteristics(params.physicalCharacteristics, params.supportedVesselTypes);

        if (!params.code) throw new Error('Code cannot be null or empty.');
        if (!params.name) throw new Error('Name cannot be null or empty.');
        if (!params.location) throw new Error('Location cannot be null or empty.');
        if (!params.physicalCharacteristics) throw new Error('Physical characteristics cannot be null.');
        if (!params.supportedVesselTypes || params.supportedVesselTypes.length === 0) {
            throw new Error('Invalid vessel types - at least one vessel type must be supported.');
        }

        this._code = params.code;
        this._name = params.name;
        this._location = params.location;
        this._physicalCharacteristics = params.physicalCharacteristics;
        this._supportedVesselTypes = params.supportedVesselTypes;
    }

    get code(): string { return this._code; }
    get name(): string { return this._name; }
    get location(): string { return this._location; }
    get physicalCharacteristics(): PhysicalCharacteristics { return this._physicalCharacteristics; }
    get supportedVesselTypes(): VesselType[] { return this._supportedVesselTypes; }

    updateName(newName: string): void {
        if (!newName) throw new Error('Name cannot be null or empty.');
        this._name = newName;
    }

    updateLocation(newLocation: string): void {
        if (!newLocation) throw new Error('Location cannot be null or empty.');
        this._location = newLocation;
    }

    updatePhysicalCharacteristics(physicalCharacteristics: PhysicalCharacteristics): void {
        this.validatePhysicalCharacteristics(physicalCharacteristics, this._supportedVesselTypes);
        this._physicalCharacteristics = physicalCharacteristics;
    }

    updateVesselTypes(newVesselTypes: VesselType[]): void {
        this.validatePhysicalCharacteristics(this._physicalCharacteristics, newVesselTypes);
        this._supportedVesselTypes = newVesselTypes;
    }

    private validatePhysicalCharacteristics(physicalCharacteristics: PhysicalCharacteristics, supportedVesselTypes: VesselType[]): void {
        if (!physicalCharacteristics) {
            throw new Error('Physical characteristics cannot be null.');
        }
        if (!supportedVesselTypes || supportedVesselTypes.length === 0) {
            throw new Error('Invalid vessel types - at least one vessel type must be supported.');
        }

        for (const vt of supportedVesselTypes) {
            if (physicalCharacteristics.length < vt.physicalCharacteristics.length) {
                throw new Error(`The dock's length is insufficient for the vessel type ${vt.name}.`);
            }
            if (physicalCharacteristics.depth < vt.physicalCharacteristics.depth) {
                throw new Error(`The dock's depth is insufficient for the vessel type ${vt.name}.`);
            }
            if (physicalCharacteristics.draft < vt.physicalCharacteristics.draft) {
                throw new Error(`The dock's max draft is insufficient for the vessel type ${vt.name}.`);
            }
        }
    }

    toDto(): DockDto {
        return {
            code: this._code,
            name: this._name,
            location: this._location,
            physicalCharacteristics: this._physicalCharacteristics,
            supportedVesselTypes: this._supportedVesselTypes.map(vt => vt.name),
        };
    }
}
